tetrominos_learning/tetrominos_learning.py

```diff
@@ -0,0 +1,291 @@
+)  # , 1.65615])
+
+# Trainable weights of the quantum network.
+with tf.name_scope('variables'):
+    r1 = tf.Variable(tf.random_normal(shape=[depth], stddev=sdev))
+    r2 = tf.Variable(tf.random_normal(shape=[depth], stddev=sdev))
+
+    theta1 = tf.Variable(tf.random_normal(shape=[depth], stddev=sdev))
+    phi1 = tf.Variable(tf.random_normal(shape=[depth], stddev=sdev))
+
+    theta2 = tf.Variable(tf.random_normal(shape=[depth], stddev=sdev))
+    phi2 = tf.Variable(tf.random_normal(shape=[depth], stddev=sdev))
+
+    sqr1 = tf.Variable(tf.random_normal(shape=[depth], stddev=sdev))
+    sqphi1 = tf.Variable(tf.random_normal(shape=[depth]))
+
+    sqr2 = tf.Variable(tf.random_normal(shape=[depth], stddev=sdev))
+    sqphi2 = tf.Variable(tf.random_normal(shape=[depth]))
+
+    dr1 = tf.Variable(tf.random_normal(shape=[depth], stddev=sdev))
+    dphi1 = tf.Variable(tf.random_normal(shape=[depth]))
+
+    dr2 = tf.Variable(tf.random_normal(shape=[depth], stddev=sdev))
+    dphi2 = tf.Variable(tf.random_normal(shape=[depth]))
+
+    kappa1 = tf.Variable(tf.random_normal(shape=[depth], stddev=sdev))
+    kappa2 = tf.Variable(tf.random_normal(shape=[depth], stddev=sdev))
+
+# List of all the weights
+parameters = [
+    r1,
+    r2,
+    theta1,
+    phi1,
+    theta2,
+    phi2,
+    sqr1,
+    sqphi1,
+    sqr2,
+    sqphi2,
+    dr1,
+    dphi1,
+    dr2,
+    dphi2,
+    kappa1,
+    kappa2,
+]
+
+# ====================================================
+#   Definition of the quantum neural network
+# ====================================================
+
+# Single quantum variational layer
+
+
+def layer(l):
+    with tf.name_scope('layer_{}'.format(l)):
+        BSgate(theta1[l], phi1[l]) | (q[0], q[1])
+        Rgate(r1[l]) | q[0]
+        Sgate(tf.clip_by_value(sqr1[l], -sq_clip, sq_clip), sqphi1[l]) | q[0]
+        Sgate(tf.clip_by_value(sqr2[l], -sq_clip, sq_clip), sqphi2[l]) | q[1]
+        BSgate(theta2[l], phi2[l]) | (q[0], q[1])
+        Rgate(r2[l]) | q[0]
+        Dgate(tf.clip_by_value(dr1[l], -disp_clip, disp_clip), dphi1[l]) | q[0]
+        Dgate(tf.clip_by_value(dr2[l], -disp_clip, disp_clip), dphi2[l]) | q[1]
+        Kgate(tf.clip_by_value(kappa1[l], -kerr_clip, kerr_clip)) | q[0]
+        Kgate(tf.clip_by_value(kappa2[l], -kerr_clip, kerr_clip)) | q[1]
+
+
+# StrawberryFields quantum simulator
+engine, q = sf.Engine(num_subsystems=modes)
+
+# Definition of the CV quantum network
+with engine:
+    # State preparation
+    Dgate(disps_alpha) | q[0]
+    Dgate(disps_beta) | q[1]
+    # Sequence of variational layers
+    for i in range(depth):
+        layer(i)
+
+# Symbolic evaluation of the output state
+state = engine.run('tf', cutoff_dim=cutoff, eval=False, batch_size=num_images)
+ket = state.ket()
+
+# Projection on the subspace of up to im_dim-1 photons for each mode.
+ket_reduced = ket[:, :im_dim, :im_dim]
+norm = tf.sqrt(tf.abs(tf.reduce_sum(tf.conj(ket_reduced) * ket_reduced, axis=[1, 2])))
+norm_inv = tf.expand_dims(
```

Comment:
Worth explaining the expand_dims part? the idea is that `norm` has `shape=[num_images]` so we make `norm_inv` have `shape=[num_images, 1, 1]`. Then, when we multiply by `ket_reduced` with `shape=[num_images, im_dim, im_dim]`, Tensorflow then uses [broadcasting](https://docs.scipy.org/doc/numpy/user/basics.broadcasting.html) rules to get `ket_processed`.

Reply:
^ The explanation can of course be less detailed than this!

Reply:
Ok I can do it. 
I was also wondering if it is better to use the `state.trace()` instead of the explicit norm formula.

Reply:
Done. I was wrong about state.trace() since we need the norm of the projected vector. However I replaced the squared norm in the regularization term at line 220 with trace.

Reply:
Nice!